### ISSUE TITLE
Error training YOLO-NAS with Custom Dataset

### ISSUE THREAD

**Issue by LZAguiar** [Deci-AI/super-gradients#1046]
### 🐛 Describe the bug

I am trying to train YOLO-NAS with a custom dataset. This dataset has about 1252 different images. However, when I try to train the network I get the following error:

```
from super_gradients.training import Trainer

trainer = Trainer(experiment_name='yolo_nas_run', ckpt_root_dir=CHECKPOINT_DIR)
trainer.train(
    model=model, 
    training_params=train_params, 
    train_loader=train_data,
    valid_loader=val_data
)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-15-e4c67dbc61e1>](https://localhost:8080/#) in <cell line: 1>()
----> 1 trainer.train(
      2     model=model,
      3     training_params=train_params,
      4     train_loader=train_data,
      5     valid_loader=train_data

5 frames
[/usr/local/lib/python3.10/dist-packages/super_gradients/training/losses/ppyolo_loss.py](https://localhost:8080/#) in _varifocal_loss(pred_logits, gt_score, label, alpha, gamma)
    892     def _varifocal_loss(pred_logits: Tensor, gt_score: Tensor, label: Tensor, alpha=0.75, gamma=2.0) -> Tensor:
    893         pred_score = pred_logits.sigmoid()
--> 894         weight = alpha * pred_score.pow(gamma) * (1 - label) + gt_score * label
    895         loss = -weight * (gt_score * torch.nn.functional.logsigmoid(pred_logits) + (1 - gt_score) * torch.nn.functional.logsigmoid(-pred_logits))
    896         return loss.sum()

RuntimeError: The size of tensor a (80) must match the size of tensor b (8) at non-singleton dimension 2
```

I have checked the data and it looks correct. I'm not sure what might be going on, but here are the training parameters, too.

```
from super_gradients.training.losses import PPYoloELoss
from super_gradients.training.metrics import DetectionMetrics_050
from super_gradients.training.models.detection_models.pp_yolo_e import PPYoloEPostPredictionCallback

train_params = {
    # ENABLING SILENT MODE
    'silent_mode': True,
    "average_best_models":True,
    "warmup_mode": "linear_epoch_step",
    "warmup_initial_lr": 1e-6,
    "lr_warmup_epochs": 3,
    "initial_lr": 5e-4,
    "lr_mode": "cosine",
    "cosine_final_lr_ratio": 0.1,
    "optimizer": "Adam",
    "optimizer_params": {"weight_decay": 0.0001},
    "zero_weight_decay_on_bias_and_bn": True,
    "ema": True,
    "ema_params": {"decay": 0.9, "decay_type": "threshold"},
    # ONLY TRAINING FOR 10 EPOCHS FOR THIS EXAMPLE NOTEBOOK
    "max_epochs": 10,
    "mixed_precision": True,
    "loss": PPYoloELoss(
        use_static_assigner=False,
        # NOTE: num_classes needs to be defined here
        num_classes=len(dataset_params['classes']),
        reg_max=16
    ),
    "valid_metrics_list": [
        DetectionMetrics_050(
            score_thres=0.1,
            top_k_predictions=300,
            # NOTE: num_classes needs to be defined here
            num_cls=len(dataset_params['classes']),
            normalize_targets=True,
            post_prediction_callback=PPYoloEPostPredictionCallback(
                score_threshold=0.01,
                nms_top_k=1000,
                max_predictions=300,
                nms_threshold=0.7
            )
        )
    ],
    "metric_to_watch": 'mAP@0.50'
}
```

### Versions

Python version: 3.10.11 (main, Apr  5 2023, 14:15:10) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.107+-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 525.85.12
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          2
On-line CPU(s) list:             0,1
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) CPU @ 2.00GHz
Stepping:                        3
CPU MHz:                         2000.140
BogoMIPS:                        4000.28
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB
L1i cache:                       32 KiB
L2 cache:                        1 MiB
L3 cache:                        38.5 MiB
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.22.4
[pip3] torch==1.13.1
[pip3] torchaudio==2.0.2+cu118
[pip3] torchdata==0.6.1
[pip3] torchmetrics==0.8.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.15.2
[pip3] torchvision==0.14.1
[pip3] triton==2.0.0
[conda] Could not collect

**Comment by wematan** [comment]
Hi,
a good guess will be a problem with the loss not configured to your dataset number of classes.
since coco has 80 classes and i guess your dataset has less (maybe 8).

**Comment by LZAguiar** [comment]
Hello,
I thought it might be something like this. However, I changed the number of classes in the training parameter ("num_classes=len(dataset_params['classes'])). Do you know if there is a need for any further modification of the parameters? (I sent them in the first comment)

**Comment by wematan** [comment]
I’ve found that in some recipes “num_classes” is configured multiple times but under different config blocks and sometimes under a different field name.

I solved my issue by changing the number of classes that the loss is configured for as well, try looking at the original recipe your trying to override to find out what the correct field name is (the recipe I used had a field named “num_cls” in the loss config block).


BR,
Matan

**Comment by LZAguiar** [comment]
I was able to solve this by also changing the number of classes when importing the "default" model. Even though the training started and trained for about 20 minutes, I got a new error:

```
[2023-05-22 19:09:52] INFO - sg_trainer.py - Using EMA with params {'decay': 0.9, 'decay_type': 'threshold'}
[2023-05-22 19:09:56] INFO - utils.py - Note: NumExpr detected 12 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
[2023-05-22 19:09:56] INFO - utils.py - NumExpr defaulting to 8 threads.
The console stream is now moved to /content/drive/MyDrive/MerchanMe/Dados/Modelos/checkpoints/yolo_nas_run/console_May22_19_09_57.txt
[2023-05-22 19:10:06] INFO - sg_trainer_utils.py - TRAINING PARAMETERS:
    - Mode:                         Single GPU
    - Number of GPUs:               1          (1 available on the machine)
    - Dataset size:                 1252       (len(train_set))
    - Batch size per GPU:           8          (batch_size)
    - Batch Accumulate:             1          (batch_accumulate)
    - Total batch size:             8          (num_gpus * batch_size)
    - Effective Batch size:         8          (num_gpus * batch_size * batch_accumulate)
    - Iterations per epoch:         156        (len(train_loader))
    - Gradient updates per epoch:   156        (len(train_loader) / batch_accumulate)

[2023-05-22 19:10:06] WARNING - sg_trainer.py - Could not set preprocessing pipeline from the validation dataset:
 .
 Before callingpredict make sure to call set_dataset_processing_params.
[2023-05-22 19:24:17] INFO - base_sg_logger.py - [CLEANUP] - Successfully stopped system monitoring process
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
[<ipython-input-20-e4c67dbc61e1>](https://localhost:8080/#) in <cell line: 1>()
----> 1 trainer.train(
      2     model=model,
      3     training_params=train_params,
      4     train_loader=train_data,
      5     valid_loader=val_data

4 frames
[/usr/local/lib/python3.10/dist-packages/super_gradients/training/transforms/transforms.py](https://localhost:8080/#) in get_equivalent_preprocessing(self)
    426 
    427     def get_equivalent_preprocessing(self) -> List:
--> 428         raise NotImplementedError
    429 
    430 

NotImplementedError:
```

**Comment by Louis-Dupont** [comment]
Hi @LZAguiar , could you please show the snippet of how you instantiate the datasets? And if there is anything you changed on these datasets after instantiation

**Comment by HissaanAli** [comment]
I found the solution ✔️

**What was happening:**
The number of classes was not set correctly, as it has to be updated at multiple places in a code. Here's the list of places I had to update it to get it to work: 

- model.get('yolo_nas_m', num_classes = YOUR_NUMBER_OF_CLASSES)
- In the "loss" field's parameters in train_params for the field called "num_classes".
- In the "valid_metrics_list" params field called "num_cls"
- In the dataset params field called for the field called "classes". This one has to be a list of all the classes rather than just a list.

It would have been a lot more easier, if we could just do it in one place !
